src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java

```diff
@@ -0,0 +1,109 @@
+    private final CaseRetentionRepository caseRetentionRepository;
+
+    public RemoveDuplicateEventsProcessorImpl(
+        @Value("${darts.events.duplicates.earliest-removable-event-date}") LocalDate earliestRemovableEventDate,
+        @Value("${darts.events.duplicates.clear-up-window}") int clearUpWindow,
+        EventRepository eventRepository,
+        CurrentTimeHelper currentTimeHelper,
+        CaseManagementRetentionRepository caseManagementRetentionRepository,
+        CaseRetentionRepository caseRetentionRepository) {
+
+        this.earliestRemovableEventDate = earliestRemovableEventDate
+            .atStartOfDay()
+            .atOffset(UTC);
+        this.clearUpWindow = clearUpWindow;
+        this.eventRepository = eventRepository;
+        this.currentTimeHelper = currentTimeHelper;
+        this.caseManagementRetentionRepository = caseManagementRetentionRepository;
+        this.caseRetentionRepository = caseRetentionRepository;
+    }
+
+    @Override
+    @Transactional
+    public void processRemoveDuplicateEvents() {
+        if (clearUpWindow < 1) {
+            throw new DartsException(
+                "The duplicate event deletion task is not configured correctly. " +
+                    "The clear-up window must be at least 1 day but is configured as " + clearUpWindow);
+        }
+        var today = currentTimeHelper.currentLocalDate();
+        var startDateTime = today.minusDays(clearUpWindow).atStartOfDay().atOffset(UTC);
+        if (startDateTime.isBefore(earliestRemovableEventDate)) {
+            startDateTime = earliestRemovableEventDate;
+        }
+        var endDateTime = today.atTime(23, 59).atOffset(UTC);
+        var duplicateEvents = new ArrayList<EventEntity>();
+
+        eventRepository.findAllBetweenDateTimesInclusive(startDateTime, endDateTime).stream()
+            .collect(groupingBy(duplicateIdentification()))
+            .forEach((key, eventList) -> {
+                eventList.sort(comparing(EventEntity::getTimestamp));
+                // Keep the last (latest) event and delete the rest
+                duplicateEvents.addAll(eventList.subList(0, eventList.size() - 1));
+            });
+
+        if (!duplicateEvents.isEmpty()) {
+            log.info("duplicate events found: {}", duplicateEvents.stream().map(EventEntity::getId).toList());
+            Spliterator<EventEntity> spliterator = duplicateEvents.spliterator();
+            List<EventEntity> chunk;
+
+            while (spliterator.estimateSize() > 0) {
+                chunk = stream(spliterator, false)
+                    .limit(CHUNK_SIZE)
+                    .collect(toList());
+
+                if (!chunk.isEmpty()) {
+                    var caseManagementIdsToBeDeleted = caseManagementRetentionRepository.getIdsForEvents(chunk);
+                    caseRetentionRepository.deleteAllByCaseManagementIdsIn(caseManagementIdsToBeDeleted);
```

Comment:
Should we add a log message indicating which events have been removed for support purposes?

Reply:
Ok, maybe move it and make clear that we are deleting these events in the message?